frontend/src/core/kernel/ui-data-registry.tsx

```diff
@@ -0,0 +1,69 @@
+/* Copyright 2023 Marimo. All rights reserved. */
+import { Logger } from "@/utils/Logger";
+import { CellId } from "../model/ids";
+import { TypedString } from "../model/typed";
+import { Objects } from "@/utils/objects";
+import { repl } from "@/utils/repl";
+
+export type DataLocatorId = TypedString<"DataLocatorId">;
+
+/**
+ * Registry to track UIElement values.
+ */
+export class UIElementDataRegistry {
+  /**
+   * Shared singleton instance.
+   */
+  static readonly INSTANCE = new UIElementDataRegistry();
+
+  private constructor() {
+    repl(this, "UI_ELEMENT_DATA_REGISTRY");
+  }
+
+  // We want fast lookups by DataLocatorId and fast removals by CellId.
+  private data = new Map<DataLocatorId, unknown>();
+  private removeHandlers = new Map<CellId, () => void>();
+
+  /**
+   * Set the datastore for a Cell.
+   */
+  setValue(
+    cellId: CellId,
+    dataStore: Record<DataLocatorId, unknown> | undefined
+  ) {
+    // If there is already a remove handler for this cell, call it to remove
+    // the old values.
+    this.removeHandlers.get(cellId)?.();
```

Comment:
Could this accidentally delete data locators that this cell doesn't "own"? I thought about it for a while and now I think the answer is "no" ... but wanted to double check with you. If I understand correctly the dataStore for a cell contains all data locators, not just ones created by this cell; that's why I originally thought that this could lead to freeing resources that you didn't want to free.